driver: adc: stm32: support dma for devices without dmamux

heinwessels opened the pull request:
Previously the STM32 DMA driver was dependent on a very specific
name for the DMA in the DTS. This hidden requirement has caused
a bit of confusion. This commit changes the driver to instead
always use the first DMA listed in the ADC node's dma property.

Also, updates the dma code in the ADC driver to support the STM32L4 (and likely more other mcus)

Fixes: #65387

heinwessels committed:
drivers: dma: stm32: add utilities to access properties by idx

Some drivers like the ADC access the DMA through it's index
and not the DMA name. This commit add some better support
for such instances.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
drivers: adc: stm32: driver now agnostic of actual dma name

Previously the STM32 DMA driver was dependent on a very specific
name for the DMA in the DTS. This hidden requirement has caused
a bit of confusion. This commit changes the driver to instead
always use the first DMA listed in the ADC node's dma property.

Should fix: #65387

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

erwango reviewed (commented):
Thanks for this work @heinwessels.

My understanding is that we're extending ADC DMA support to 'non dmamux' compatible series.

Would you mind adding at least target of this kind to tests/drivers/adc/adc_dma ?

heinwessels commented:
> My understanding is that we're extending ADC DMA support to 'non dmamux' compatible series.

Yes, that's the goal.

> Would you mind adding at least target of this kind to tests/drivers/adc/adc_dma ?

I would like to, but it doesn't seem to work fully just yet. I found a possible problem which I mentioned on the Discord in the `#adc` channel. If I can verify that it does work I will add the test.

My plan was to merge this PR first as it works standalone IMO. It doesn't imply functionality without a dmamux. And once functionality is verified create a new PR with test (and maybe a sample) that has the title something explicit like "support non dmamux mcus".

Would you prefer I put everything into one PR?

heinwessels committed:
drivers: adc: stm32: add dma support for other mcus

This commit adds support for more STM32 CPUs that has
a different DMA interface. This was tested only for
the nucleo_l476rg.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
tests: drivers: adc_dma: add test for nucleo_l476rg

Add test to verify the ADC + DMA functionality works
for the STM32L476 MCU.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
@erwango I've added tests for the `nucleo_l476rg`. I can't run it though as I don't have this board. The person who brought up this issue initially said they might get time to run it on their hardware later this week.

erwango commented on tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay:
```diff
@@ -0,0 +1,18 @@
+ */
+
+&adc1 {
+	dmas = < &dma1 1 0 (STM32_DMA_MODE_NORMAL | STM32_DMA_PERIPH_TO_MEMORY |
+		STM32_DMA_MEM_INC |  STM32_DMA_MEM_16BITS | STM32_DMA_PERIPH_16BITS) >;
```
```suggestion
	dmas = < &dma1 1 0 (STM32_DMA_PERIPH_RX |  STM32_DMA_MEM_16BITS | STM32_DMA_PERIPH_16BITS) >;
```

heinwessels replied:
Done

erwango commented on tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_ADC_STM32_DMA=y
+CONFIG_ADC_ASYNC=y
```
This is not strictly related to this PR but I wonder if these 2 options should stay independent.
Can we have CONFIG_ADC_ASYNC=y and CONFIG_ADC_STM32_DMA=n ? or the other way round ?

heinwessels replied:
> Can we have CONFIG_ADC_ASYNC=y and CONFIG_ADC_STM32_DMA=n ?

I believe this is possible, and that these are independent. However, the reason `CONFIG_ADC_ASYNC=y` is because there's an async test part of this suite.

erwango reviewed (dismissed):
One pending question, but otherwise LGTM

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -216,6 +216,35 @@ static bool init_irq = true;
+#else /* defined(CONFIG_SOC_SERIES_STM32H7X) */
+
+	/* Default mechanism for other MCUs */
+	LL_ADC_REG_SetDMATransfer(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
```
A quick look through the LL shows that the macro `LL_ADC_REG_DMA_TRANSFER_LIMITED` does not exist for F1 and F37x.
For those, the choice is between `LL_ADC_REG_DMA_TRANSFER_NONE` or `LL_ADC_REG_DMA_TRANSFER_UNLIMITED`.
A `#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc)` should be added for these cases. What we do inside is up to you: unlimited seems to imply a circular buffer, which may be hard to add, so maybe we can just not support this case for now and return an error.

heinwessels replied:
I would prefer to error as "not supported", as I do not have hardware or time to investigate all potential edge cases at the moment. And my idea was that my code as-is doesn't compile then it already implies "not supported", but I can make this explicit.

heinwessels replied:
Added an explicit error

gautierg-st reviewed (changes requested):
Need to add a special case for F1 compatibles, otherwise looks good

gautierg-st reviewed (approved):
Thanks
